### DISCUSSION TITLE
Discussion on Zephyr's Poor Real-Time Performance

### DISCUSSION THREAD

**Discussion by hlleng** [zephyrproject-rtos/zephyr#79785] [General] [answered]
Recently, the well-known embedded systems expert Jacob Beningo wrote the '2024 Real-Time Operating System Performance Report.' The report tested several commonly used RTOSes, including PX5, ThreadX, Zephyr, and FreeRTOS, to evaluate real-time performance through test cases such as coroutine scheduling, preemptive scheduling, memory allocation, message handling, and synchronization processing. Based on the results, Zephyr's real-time performance was nearly the worst. What could be the reason for this?

![image](https://github.com/user-attachments/assets/a1539fcf-0550-43b1-b8e7-b3b03177fe2e)
[Beningo-RtosPerformance-2024-Report.pdf](https://github.com/user-attachments/files/17348393/Beningo-RtosPerformance-2024-Report.pdf)

**Comment by cocoeoli** [comment]
From this result, Zephyr still has a lot of space for performance optimization.

**Comment by heinwessels** [comment]
You should read the follow up article by the same person: https://www.beningo.com/is-rtos-performance-affected-by-an-mpu/#

Essentially, in the initial benchmark you posted Zephyr had a Memory Protection Unit enabled (MPU) enabled, which has an performance impact because it adds safety features.

Quote from the article:
> Zephyr jumped from a relative performance against PX5 RTOS of 42% to 70.1%! That’s a 28% improvement in the benchmark and a nearly 67% improvement against itself! All because the MPU and a few additional checks were disabled!
You might argue that disabling the MPU in the benchmark is cheating, but since the other RTOSes in the study did not have them enabled, it really levels the playing field and makes it possible to compare the RTOSes more accurately.

**Reply by aniltirli** [reply]
As mentioned in that article Zephyr's defaults are really MPU, asserttions and error-checking enabled ? I am not sure about that, in my understanding from the documentation, i think you need to set those in config file to enable them, am i wrong ?

**Comment by nashif** [comment]
quick update as things improve in zephyr and some of the issues are being addressed. Here is the state of things:

![image](https://github.com/user-attachments/assets/98518201-d23b-4a33-b8c3-7b11d528d112)


![image](https://github.com/user-attachments/assets/646a3c47-cd28-4d92-9d92-8a4ca4cbb4dd)



++ with changes currently in review and pending merge.

**Reply by DBS06** [reply]
@nashif Hi, is it possible to have access to the test code which you are using?
I am doing right now a similar benchmarks for different RTOSes and I am curious to see the configuration and the test code.

**Reply by henrikbrixandersen** [reply]
@DBS06 Have you seen https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/benchmarks/thread_metric ?

**Reply by DBS06** [reply]
> @DBS06 Have you seen https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/benchmarks/thread_metric ?

Thanks a lot!!!
Actually I am aware of zephyr/tests/benchmarks/ but I wasn't aware of this specific one. I started my benchmark-project 3 months ago and had to stop it because other things were more urgent.
So thanks a lot!

**Comment by nashif** [comment, marked as answer]
Update just before feature freeze of 4.1..... (c8f640a61034e57c9eec186e1e6d396b2f8322a0)

![image](https://github.com/user-attachments/assets/4e8c8b2e-8c2c-4e8c-89e0-1aa865f5befe)
![image](https://github.com/user-attachments/assets/743b8cda-692a-4b46-a83a-339b1471dbcb)


Few changes currently in review would improve things even more, stay tuned.

**Comment by romkell** [comment]
Also if that discussion is a bit dated:

Do we also have figures on the kernels worst case IRQ lock-out times?

ARM Cortex-M >= M3 do have an interrupt controller (NVIC) with ISR level lock register (BASEPRI) allowing for CONFIG_ZERO_LATENCY_IRQS.
ARM Cortex-M <= M1 and other architectures do not have that feature to avoid the kernel locking critical ISRs introducing some jitter to their timing.

Do we know those figures as clock cycles related to the gcc version and instruction set or probably simpler as time on certain SOC derivatives e.g. ARM Cortex-M0+ or others @ defined clock speeds and gcc version?
